File: internal/locks/locks.go
package locks

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

type LockFile struct {
	SourcePath  string                `json:"source_path"`
	SourceHash  string                `json:"source_hash"`
	ContextHash string                `json:"context_hash,omitempty"`
	Outputs     map[string]OutputLock `json:"outputs"`
	UpdatedAt   string                `json:"updated_at"`
}

type OutputLock struct {
	Path        string `json:"path"`
	Hash        string `json:"hash"`
	ContextHash string `json:"context_hash,omitempty"`
	CheckedAt   string `json:"checked_at"`
}

func LockPath(root, sourcePath string) string {
	return filepath.Join(root, ".l10n", "locks", sourcePath+".lock")
}

func Read(root, sourcePath string) (*LockFile, error) {
	path := LockPath(root, sourcePath)
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var lock LockFile
	if err := json.Unmarshal(data, &lock); err != nil {
		return nil, err
	}
	return &lock, nil
}

func Write(root, sourcePath string, lock LockFile) error {
	lock.UpdatedAt = time.Now().UTC().Format(time.RFC3339)
	path := LockPath(root, sourcePath)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(lock, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

File: src/locks.ts
import { join, dirname } from "path";
import { mkdir, readFile, writeFile } from "fs/promises";

export interface OutputLock {
  path: string;
  hash: string;
  context_hash?: string;
  checked_at: string;
}

export interface LockFile {
  source_path: string;
  source_hash: string;
  context_hash?: string;
  outputs: Record<string, OutputLock>;
  updated_at: string;
}

export function lockPath(root: string, sourcePath: string): string {
  return join(root, ".l10n", "locks", sourcePath + ".lock");
}

export async function readLock(
  root: string,
  sourcePath: string,
): Promise<LockFile | null> {
  const path = lockPath(root, sourcePath);
  try {
    const data = await readFile(path, "utf-8");
    return JSON.parse(data) as LockFile;
  } catch (err: any) {
    if (err.code === "ENOENT") return null;
    throw err;
  }
}

export async function writeLock(
  root: string,
  sourcePath: string,
  lock: LockFile,
): Promise<void> {
  lock.updated_at = new Date().toISOString();
  const path = lockPath(root, sourcePath);
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, JSON.stringify(lock, null, 2) + "\n");
}
